wgpu-core/src/registry.rs

```diff
@@ -0,0 +1,537 @@
+        match self.get(id) {
+            Ok(resource) => Some((resource, resource.life_guard(), resource.device_id())),
+            Err(InvalidId::ResourceInError { .. }) => {
+                self.unregister(id).ok()?;
+                None
+            }
+            Err(e) => {
+                log::error!("Tried to drop invalid {} id: {}", T::TYPE, e);
+                None
+            }
+        }
+    }
+
+    pub fn insert_error(&self, id: T::Id, implicit_failure: &'static str) {
+        let (index, epoch, _) = id.unzip();
+        unsafe {
+            self.storage
+                .fill(index, epoch, Err(Cow::Borrowed(implicit_failure)))
+        }
+    }
+
+    pub fn force_replace(&self, id: T::Id, value: T) {
+        let (index, epoch, _) = id.unzip();
+        unsafe { self.storage.overwrite(index, epoch, value) }
+    }
+
+    pub fn label_for_resource(&self, id: T::Id) -> String {
+        match self.get(id) {
+            Ok(resource) => resource.label().to_owned(),
+            Err(_) => format!("{:?}", id),
+        }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (T::Id, &T)> {
+        IteratorDataAdapter {
+            _data: self.identity_manager.lock(),
+            // SAFETY: The identity manager lock is taken and will be kept
+            // alive as long as the iterator is alive.
+            iter: unsafe { self.storage.iter(A::VARIANT) },
+        }
+    }
+
+    /// # Safety:
+    ///
+    /// This function must be called _as if_ it takes a &mut self argument.
+    ///
+    /// If this is called with anything other than exclusive access to the registry,
+    /// all bets are off.
+    pub unsafe fn iter_mut(&self) -> impl Iterator<Item = (T::Id, &mut T)> {
+        self.storage.iter_mut(A::VARIANT)
+    }
+
+    /// # Safety:
+    ///
+    /// This function must be called _as if_ it takes a &mut self argument.
+    ///
+    /// If this is called with anything other than exclusive access to the registry,
+    /// all bets are off.
+    pub unsafe fn remove_all(&self) -> impl Iterator<Item = T> + '_ {
+        self.storage.remove_all(A::VARIANT)
+    }
+
+    pub fn generate_report(&self) -> hub::StorageReport {
+        todo!()
+    }
+}
+
+impl<A, T, F> std::ops::Index<id::Valid<T::Id>> for Registry<A, T, F>
+where
+    A: hub::HalApi,
+    T: hub::Resource,
+    F: hub::IdentityHandlerFactory<T::Id>,
+{
+    type Output = T;
+
+    fn index(&self, id: id::Valid<T::Id>) -> &Self::Output {
+        self.get(id.0).unwrap()
+    }
+}
+
+struct IteratorDataAdapter<D, I> {
+    _data: D,
+    iter: I,
+}
+
+impl<D, I> Iterator for IteratorDataAdapter<D, I>
+where
+    I: Iterator,
+{
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next()
+    }
+}
+
+const STORAGE_BLOCK_SIZE: usize = 32;
+
+pub struct Storage<T> {
+    blocks: [UnsafeCell<Option<Box<StorageBlock<STORAGE_BLOCK_SIZE, T>>>>; 256],
```

Comment:
That would work, would be safer and would be easy to change back if it's an issue in arm

Reply:
Great!

Somehow `Storage` feels a lot like `ArrayVec<T, STORAGE_BLOCK_SIZE>` for some appropriately fancy `T`. So maybe we can side-step this `UnsafeCell` somehow as well. Letting the original `Storage`'s elements be enums wasn't a major perf problem, right?

Reply:
I don't think we can use ArrayVec directly because it needs a &mut to add an element and we need to add elements on &mut

As for the enum, the reason the storage isn't online is because the enum is going to become a fancy atomic at some point so we wouldn't need the lock at all. Even if we kept it as a lock, it's much easier to avoid UB nonsense where we don't need to go _through_ a mutex to get unlocked access to something.